File: LoadMasterService/src/main/java/fu/se184491/loadmaster_be/entity/optimization/PackagePlacement.java
package fu.se184491.loadmaster_be.entity.optimization;

import fu.se184491.loadmaster_be.entity.optimization.LoadPlan;
import fu.se184491.loadmaster_be.entity.cargo.CargoPackage;


import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.*;

@Entity
@Table(name = "package_placements", uniqueConstraints = @UniqueConstraint(
        name = "uk_placement_plan_package", columnNames = {"load_plan_id", "package_id"}))
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class PackagePlacement {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "load_plan_id")
    private LoadPlan loadPlan;

    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "package_id")
    private CargoPackage cargoPackage;

    @Column(name = "pos_x") private Integer posX;
    @Column(name = "pos_y") private Integer posY;
    @Column(name = "pos_z") private Integer posZ;
    @Column(name = "packed_length") private Integer packedLength;
    @Column(name = "packed_width") private Integer packedWidth;
    @Column(name = "packed_height") private Integer packedHeight;

    @Min(0) @Max(5) @Column(name = "rotation_type")
    private Integer rotationType;

    @Column(name = "step_sequence")
    private Integer stepSequence;

    @Builder.Default @Column(name = "pinned")
    private Boolean pinned = false;
}

File: OptimizeService/app/entity/optimization/package_placement.py
from sqlalchemy import Column, Integer, BigInteger, ForeignKey, Numeric, Boolean
from sqlalchemy.orm import relationship

from app.config.database import Base


class PackagePlacement(Base):
    __tablename__ = "package_placements"

    id = Column(Integer().with_variant(BigInteger, "postgresql"), primary_key=True, autoincrement=True)
    load_plan_id = Column(Integer().with_variant(BigInteger, "postgresql"), ForeignKey("load_plans.id"), nullable=False)
    package_id = Column(Integer().with_variant(BigInteger, "postgresql"), ForeignKey("packages.id"), nullable=True)
    pos_x = Column(Numeric(10, 2), nullable=True)
    pos_y = Column(Numeric(10, 2), nullable=True)
    pos_z = Column(Numeric(10, 2), nullable=True)
    packed_length = Column(Numeric(10, 2), nullable=True)
    packed_width = Column(Numeric(10, 2), nullable=True)
    packed_height = Column(Numeric(10, 2), nullable=True)
    rotation_type = Column(Integer, nullable=True, default=0)
    step_sequence = Column(Integer, nullable=True, default=1)
    pinned = Column(Boolean, default=False)

    load_plan = relationship("LoadPlan", back_populates="placements")
    cargo_package = relationship("Package")

    @property
    def loading_sequence(self) -> int:
        return self.step_sequence or 1

    @loading_sequence.setter
    def loading_sequence(self, val: int) -> None:
        self.step_sequence = val
